modules/client-common/src/main/java/org/apache/ignite/client/proto/query/QueryEventHandlerImpl.java

```diff
@@ -0,0 +1,211 @@
+                "Failed to fetch results for cursor id " + req.cursorId() + ". Error message: " + ex.getMessage());
+        }
+
+        return new JdbcQueryFetchResult(fetch, hasNext);
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcBatchExecuteResult batch(JdbcBatchExecuteRequest req) {
+        return new JdbcBatchExecuteResult(UNSUPPORTED_OPERATION,
+            "ExecuteBatch operation is not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override public JdbcQueryCloseResult close(JdbcQueryCloseRequest req) {
+        Cursor<List<?>> cur = state.remove(req.cursorId());
+
+        if (cur == null)
+            return new JdbcQueryCloseResult(JdbcResponse.STATUS_FAILED,
+                "Failed to find query cursor with ID: " + req.cursorId());
+
+        try {
+            cur.close();
+        }
+        catch (Exception ex) {
+            state.remove(req.cursorId());
+
+            return new JdbcQueryCloseResult(JdbcResponse.STATUS_FAILED,
+                "Failed to close SQL query [curId=" + req.cursorId() + "]. Error message: " + ex.getMessage());
+        }
+
+        return new JdbcQueryCloseResult();
+    }
+
+    /**
+     * Creates jdbc result for the cursor.
+     *
+     * @param cur Sql cursor for query.
+     * @param req Execution request.
+     * @return JdbcQuerySingleResult filled with first batch of data.
+     */
+    private JdbcQuerySingleResult createJdbcResult(SqlCursor<List<?>> cur, JdbcQueryExecuteRequest req) {
+        long cursorId = CURSOR_ID_GENERATOR.getAndIncrement();
+
+        state.put(cursorId, cur);
+
+        List<List<Object>> fetch = fetchNext(req.pageSize(), cur);
+        boolean hasNext = cur.hasNext();
+
+        switch (cur.getQueryType()) {
+            case QUERY:
+                return new JdbcQuerySingleResult(cursorId, fetch, !hasNext);
+            case DML:
+            case DDL: {
+                if (!validateDmlResult(fetch, hasNext))
+                    return new JdbcQuerySingleResult(JdbcResponse.STATUS_FAILED,
+                        "Unexpected result for DML query [" + req.sqlQuery() + "].");
+
+                return new JdbcQuerySingleResult(cursorId, (Long)fetch.get(0).get(0));
+            }
+            case FRAGMENT:
+            case EXPLAIN:
+            default:
+                return new JdbcQuerySingleResult(UNSUPPORTED_OPERATION,
+                    "Query type [" + cur.getQueryType() + "] is not supported yet.");
+        }
+    }
+
+    /**
+     * Validate dml result. Check if it stores only one value of Long type.
+     *
+     * @param fetch Fetched data from cursor.
+     * @param next HasNext flag.
+     * @return Boolean value indicates if data is valid or not.
+     */
+    private boolean validateDmlResult(List<List<Object>> fetch, boolean next) {
+        if (next)
+            return false;
+
+        if (fetch.size() != 1)
+            return false;
+
+        if (fetch.get(0).size() != 1)
```

Comment:
is it possible to have IndexOutOfBoundsException in case fetch.size() is 0?

Reply:
We have already checked fetch to have only one item. (lines 187\188)